### PR TITLE
4985 breadcrumbs

### PR DESCRIPTION
**Pull request by lzach83**
## Link to Issue
Closes: #4985 

## Description of Changes

- Added breadcrumbs to all pages
- Stand alone pages have breadcrumbs in it's active state, should have hover state but not be clickable. 
- Parent pages hover state should be clickable when clicked, takes user to that page.

### CODE REVIEW

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:8]: this would be good as a "container" holding the logic, but we already have `CWBreadcrumbs` component, which is used btw in some sites, so currently in those sites there are 2 breadcrumbs displayed which is not good. 

image

- you should take CWBreadcrumbs component and adjust it into new styling
- you should remove usage of CWBreadcrumbs from all sites
- you should use CWBreadcrumbs in you Breadcrumbs component to display the new breadcrumbs

Basically `Breadcrumbs` should be container which handles the logic and `CWBreadcrumbs` should be a dumb component that only displays the breadcrumbs passed as a prop.

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:49]: dont use `x` or other non meaningfully names for variables

**Review by masvelio — COMMENTED** [review]
These are the requirements from the ticket in bold followed by my comments:
- [ ] **Left padding should mirror left padding (or margin) of page content, as this varies across the site**. - you should probably go site by site and check it because right now it seems to be not handled everywhere
image
image
image

- [ ] **When breadcrumbs overflow, horizontal scrolling behavior is activated.** - now no horizontal scrolling but the title goes multiline
- [ ]   **On mobile, or screens 600px and under, breadcrumbs are no longer sticky. View prototype below for more.** - it stays fixed
image

- [x] if we can't click the breadcrumb, we should change cursor to normal, not pointer
- [x] **Breadcrumb copy mirrors page header copy.** - now breadcrumbs are derived from the url, so first thing is that they do not always sync page header, and second, the should not use dashes or camelCase for example:
image
image

- [ ] according to ACs, explore page is not a section, it is standalone page
image

- [x] I am pretty sure that we do not want to display % in the topic names, and the thread id prefix in the title names 
image
image

- [ ] first thing is that we should not display community id in the breadcrumbs. Second, that the discussion is a section, not the community (if you remove community id it would be ok) 
image

- [ ] It looks like the sections are messed up - For governance and discussion, the tooltip copy should be `this is an app, not a selectable page`
image

image

**Comment by masvelio** [comment]
This PR has quite a few loopholes, both in the code and in the UX. @lzach83 please take your time to understand the requirements written in the ticket and figma as they are very well explained by @sachben91 and @jessmart1213. Do not hesitate to ask if something is not clear in the design or code.

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (jnaviask) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:34]: cc @gdjohnson to comment on our ideal inline documentation style

**Comment by lzach83** [comment]
Still to be done 

- Explore and implement not relying on the url to build the breadcrumbs, maybe utilize the `useMatches` from react-router-dom, or some sort of prebuilt dictionary 
- On scroll when in mobile, have the breadcrumb stay with the header instead of remaining sticky
- Fix a few more of the tooltips from displaying incorrect information.

**Comment by lzach83** [comment]
Everything is good to go except for: 

- On mobile, or screens 600px and under, breadcrumbs are no longer sticky. View prototype below for more.

I have to explore how to track scroll to nonstick the breadcrumb, unless another solution strikes me.

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/data.ts
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/Breadcrumbs/data.ts:50]: the topics cannot be hardcoded because those are dynamic and differ between communities

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/Sublayout.tsx:67]: could you use some css to do that? You can use `position: sticky` or `position: fixed` for desktop resolution and using media query you can use static position in mobile resolution

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (masvelio) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:5]: ?

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (masvelio) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:10]: ?

**Review by masvelio — CHANGES_REQUESTED** [review]
1. basically non of the todo list items I have pointed out in last review is resolved. 
2. most of my comments from previous review are not addressed in the code and there is no acknowledgement from you site:
   - either you agree with the comment, commit the fix and report back in the comment
   - or you do not agree with the comment and you reply back why you think this should be done in other way

3. I still see a lot of bugs, even more than in previous review eg

**Text in tooltips are not centered**
image

**Some breadcrumbs points to dead pages**

https://github.com/hicommonwealth/commonwealth/assets/14819225/39d6b44a-495c-4dd6-b41f-fa956d36d6d7

https://github.com/hicommonwealth/commonwealth/assets/14819225/0a3f9c76-eb7c-424f-8c83-126d755f6552


**Breadcrumb on search does not work properly**

https://github.com/hicommonwealth/commonwealth/assets/14819225/9775f9ee-2d48-40ae-8859-bc63e148d2a0

**Breadcrumbs in some pages are still not aligned to the left **

image

image


And of course all other things from my previous review, refer to the todo list from [this comment](https://github.com/hicommonwealth/commonwealth/pull/5093#pullrequestreview-1630820650)

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:5]: This is finished and will be in the next commit. (Making this note to keep progress)

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx]: This has been finished. (Making this note to keep progress)

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx]: This has been finished. (Making this note to keep progress)

File: packages/commonwealth/client/styles/components/Breadcrumbs/Breadcrumbs.scss
Comment (lzach83) [reply packages/commonwealth/client/styles/components/Breadcrumbs/Breadcrumbs.scss]: This has been finished. (Making this note to keep progress)

File: packages/commonwealth/client/styles/components/Breadcrumbs/Breadcrumbs.scss
Comment (lzach83) [reply packages/commonwealth/client/styles/components/Breadcrumbs/Breadcrumbs.scss:3]: This has been finished and will be in the next commit. (Making this note to keep progress)

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx]: This has been finished. (Making this note to keep progress)

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx]: Went a different direction on this entirely. But it's finished. (Making this note to keep progress)

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:49]: @gdjohnson Can we get an update on this?

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx]: This has been finished. (Making this note to keep progress)

File: packages/commonwealth/client/styles/components/Breadcrumbs/Breadcrumbs.scss
Comment (lzach83) [reply packages/commonwealth/client/styles/components/Breadcrumbs/Breadcrumbs.scss:5]: This has been finished and will be in the next commit. (Making this note to keep progress)

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/data.ts
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/data.ts:51]: Spoke about this async but for public knowledge I was doing some local testing and it made it into my commit.

**Comment by lzach83** [comment]
I committed but never pushed my previous commit so there was a little bit of confusion during the previous CR.

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/Sublayout.tsx:67]: I misunderstood the AC's/Reqs. Spoke to product team and they cleared this up.

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (gdjohnson) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:48]: Had a meeting today with Forest about JSDoc, but will be a day or two before I have a tentative style guide for us.

My recommendation is we merge as-is and I will circle back & follow up (and if necessary, make any style adjustments).

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:10]: Resolved this in most recent commit.

**Comment by lzach83** [comment]
Good to review.

**Comment by masvelio** [comment]
> Good to review.

will take a look on this tomorrow morning

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (masvelio) [reply packages/commonwealth/client/scripts/views/Sublayout.tsx:67]: 1. long breadcrumbs should have horizontal scroll
image

2. in general, breadcrumbs should not cover the body of the page. in several cases, breadcrumb covers the header
image

**Comment by lzach83** [comment]
Was talking with @masvelio  about my most recent iteration on this ticket, I took a route of implementing the Breadcrumbs that is seemingly not going to work. I build a component and put it in the body in the `<Sublayout/>`  the issue is, padding/margin is not unified across the app. So pages where the header is centered like: /createCommunity/starter, design wants the breadcrumbs offset to the left of the centered header.. which is extremely difficult or impossible due the structure of the app/pages without adding a bunch more breakpoints or adding the `<Breadcrumbs />`to each individual page.

I am gathering all of the pages now.

cc: @jnaviask

**Comment by masvelio** [comment]
Per @lzach83  request I clicked around and spotted few issues. I shared them with @lzach83 but it seems that there is some sidebar bug that affects the breadcrumbs, so I will re-review this again

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:48]: Sounds good, i will update once #5241 is fixed

Comment (masvelio) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:48]: thanks for you input @gdjohnson!

**Comment by lzach83** [comment]
Removed the use of `isMobile` because it does not track resizing windows.

**Comment by lzach83** [comment]
Brief loom on functionality: https://www.loom.com/share/1cd46f171fd041d7bff20e858c702298?sid=aa4d4b08-7ed1-4ce2-b39e-d12bf8278def

**Comment by dillchen** [comment]
image

On scroll up? breadcrumbs are staying static, it breaks a bit of our UX

**Comment by dillchen** [comment]
Similar behavior on reaching the bottom of the page:

image

**Comment by dillchen** [comment]
image

Some breadcrumbs (while on mobile still on mobile lines)

**Comment by dillchen** [comment]
As a Note. cc @jessmart1213 @zakhap, I believe we should use ellipsis rather than a horizontal scroll for any breadcrumb section name longer than 40 chars or screen width.

This is the behavior in Notion:

image

I logged a similar issue #5272 as well which I believe should have a similar solution.

I think this is important because breadcrumbs a dramatic improvement to mobile nav, and this would be the final touch!

**Review by dillchen — CHANGES_REQUESTED** [review]
Mostly visual bugs that are noted in comments, it really is the home stretch!

**Comment by zakhap** [comment]
> As a Note. cc @jessmart1213 @zakhap, I believe we should use ellipsis rather than a horizontal scroll or multi-line approach for any breadcrumb section name longer than 40 chars or screen width.
> 

I think simply truncating breadcrumb sections that are too long is fine to this PR (how many characters?)

I do not think adding the ellipses and the dropdown menu showing the rest of the hierarchy is in scope for this work as we do not yet have pages nested that deep to require such an improvement. This would also be a new DS Pattern.

**Comment by dillchen** [comment]
adding some context: I am all good with @zakhap re truncation:

Outside of that:

> I do not think adding the ellipses and the dropdown menu showing the rest of the hierarchy is in scope for this work as we do not yet have pages nested that deep to require such an improvement. This would also be a new DS Pattern.

agreed, that slipped my mind and was out of scope

File: packages/commonwealth/client/styles/index.scss
Comment (lzach83) [inline packages/commonwealth/client/styles/index.scss:13]: This was added in order to prevent over-scrolling mainly used on Mac. The more long term solution standardized padding/margin across the app which is in the roadmap.

**Review by dillchen — APPROVED** [review]

**Comment by masvelio** [comment]
@lzach83 I am highlighting most important issues I run into:

- clicking breadcrumb in proposals and snapshots leads to wrong / dead page

https://github.com/hicommonwealth/commonwealth/assets/14819225/2699b01c-d63f-4f84-b218-3cc9c2b0f79a

https://github.com/hicommonwealth/commonwealth/assets/14819225/6cd24353-e7bf-4c80-a2f7-3b83e84de84a

- community prefix is visible as a first breadcrumb despite it is not visible on the figma designs
- apart from that, sometimes the page titles do not correspond with breadcrumbs
- there are some places where the top-bottom paddings are too big/too small

I think that before merging this, we should deploy it to the demo environment and ask design/product to click around if they are ok how breadcrumbs look and work.

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/Sublayout.tsx:53]: what is the reason that this piece is both here and in other useEffect below? I think this should be only in one place, when the window is getting resized?

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/Sublayout.tsx:29]: why 425 exactly? Also, we have bunch of utils that handles the breakpoints and check innerWidth of the window, especially, that it is enclosed in nice hook `useBrowserWindow`. We should not add custom code that repeats the functionality of created utils

File: packages/commonwealth/client/scripts/views/pages/discussions/DiscussionsPage.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/pages/discussions/DiscussionsPage.tsx:74]: again, this 425 should be either defined as a global var or better, this logic should be handled by the utils functions I mentioned above

File: packages/commonwealth/client/scripts/views/pages/search/index.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/pages/search/index.tsx:74]: same here, Do Not Repeat your code => this should be util/hook

**Comment by dillchen** [comment]
https://www.loom.com/share/571a8fffbb1c4a8086da2bc8b5409b50

during an informal review with @jessmart1213 we discovered this, on mobile breadcrumb padding should stay consistent.

----
Second cc: @zakhap / @jessmart1213 it also came to attention that the level of nesting on scoped pages may not be correct-- I.e we should not show communities. This is something I missed as part of the specification on the issue. I think that as things are moving to final design + product review, and as @lzach83 gets back from OOO we should do a synchronous review early next week

**Comment by CowMuon** [comment]
Re-requested review from Dillon just to clear that approval, since this PR is still in draft.

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/Sublayout.tsx]: Wrapped these up

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/Sublayout.tsx]: This will be addressed in the newest commit.

File: packages/commonwealth/client/scripts/views/pages/discussions/DiscussionsPage.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/pages/discussions/DiscussionsPage.tsx]: Fixed in next commit

File: packages/commonwealth/client/scripts/views/pages/search/index.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/pages/search/index.tsx]: Fixed in next commit

**Comment by lzach83** [comment]
All caught up on the comments, I ran into an error with proposals locally not working giving me a `Cannot destructure property 'lastCommentedOn' of 'thread' as it is undefined.` error. 

When eng is back from OOO -- we should meet on this.

**Comment by jessmart1213** [comment]
Responding to an earlier comment [here](https://github.com/hicommonwealth/commonwealth/pull/5093#issuecomment-1773001214):

### Re: padding

> during an informal review with @jessmart1213 we discovered this, on mobile breadcrumb padding should stay consistent.

On both desktop and mobile, the margin on the breadcrumbs should match that of page. Currently our pages have various page margins and lack consistency. 

_Future note:  Ideally all pages on Desktop would have a 56px margin and on mobile screens a 16px margin, but this expands the scope of this project and gets into grids and layouts which we have yet to include in the design system._

### Re: breadcrumb nesting

> Second cc: @zakhap / @jessmart1213 it also came to attention that the level of nesting on scoped pages may not be correct-- I.e we should not show communities. This is something I missed as part of the specification on the issue. I think that as things are moving to final design + product review--that said I think it was nice to feel / see 3 levels on scoped pages, and as @lzach83 gets back from OOO we should do a synchronous review early next week to discuss if the implemented scoping is okay

Screen Shot 2023-11-27 at 4 40 19 PM

Notice in this screen, that Osmosis is the parent element, which does not follow the design. The design states that the parent element on this All Discussions page would be the **Parent Section** "Discussion". 

This can be a bit confusing as our breadcrumbs do not follow convention, because our site is unconventional. Lmk if you want a walkthrough or more info. Also definitions of parent section can be found[ here ](https://github.com/hicommonwealth/commonwealth/issues/4985#issue-1884912255)

File: packages/commonwealth/client/scripts/views/pages/search/index.tsx
Comment (lzach83) [inline packages/commonwealth/client/scripts/views/pages/search/index.tsx:60]: I have a question out to @jessmart1213 on how to handle this functionality, there is a hook made `useWindowResize` but this may not even be needed if breadcrumbs won't exist on this niche page. It's not currently on the figma's sitemap.

**Comment by dillchen** [comment]
Noting Visual Bugs:

- [ ] Dashboard

image

**Review by dillchen — CHANGES_REQUESTED** [review]
Fix above comment

**Comment by lzach83** [comment]
> Noting Visual Bugs:
> 
> * [x]  Dashboard, right sidebar Suggested communities is cut off
> 
> image
> * [x]  Profile Page (Site Map may be incorrect?)
> 
> image
> * [ ]  Profile Page (Site Map and Back Button navigation not working as expected
> 
> https://www.loom.com/share/9c0889410f4b4796a9e08a2883b3efac?from_recorder=1&focus_title=1
> 
> **Out of scope**
> 
> but the left alignment and top alignment of content and breadcrumbs bothers me, but will be out of scope until #6094 gets merged in probably
> 
> See example:
> 
> https://www.loom.com/share/15fcd759a1b944329b2cb619d63c6f83?from_recorder=1&focus_title=1

Made a bug ticket for the profile bug since it is unrelated to this PR: #6201 

Fixed the other two issues.

**Review by dillchen — APPROVED** [review]
LGTM

File: packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/components/Breadcrumbs/Breadcrumbs.tsx:11]: Screenshot

Putting a screenshot here just to document that this is done incase we have to come back, just for my own sanity.

File: packages/commonwealth/client/scripts/views/pages/search/index.tsx
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/pages/search/index.tsx]: @jessmart1213 And I spoke about the above and we are going to leave it as-is for now. Design and product are going to revamp the search page and that will eventually surface more clarity on what this will look like.

**Review by Miaplacidus — APPROVED** [review]
Breadcrumbs turned out to be pretty complex! 😅
Well done, Mr. Lee!

**Comment by masvelio** [comment]
https://github.com/hicommonwealth/commonwealth/assets/14819225/95e5ce8f-27b0-476b-886b-109a82d2e613

1. When I go to governance / snapshots and select snapshot page, then the breadcrumbs are messed up + on click there is an error page
2. should breadcrumb be blue on hover? cant see that on designs

Other than that I did not run into any other issues. Now will take a look on the code 👀

**Review by masvelio — APPROVED** [review]
left some comments, approving in advance

**Comment by dillchen** [comment]
Let’s get this baby in

Sent via Superhuman iOS ***@***.***>

**Pull request merged**
